Meadowlark-js/packages/meadowlark-core/src/middleware/AuthorizationMiddleware.ts

```diff
@@ -3,36 +3,241 @@
+      Logger.debug(
+        `${moduleName}.authorize OAuth server responded that Meadowlark own token is not valid`,
+        frontendRequest.traceId,
+      );
+
+      // Re-fetch Meadowlark own token - in case it expired - and try one more time
+      const fetchOwnAccessTokenResult: RequestOwnAccessTokenResult = await requestOwnAccessToken(frontendRequest);
+
+      if (fetchOwnAccessTokenResult.success) {
+        // Cache the new Meadowlark token
+        cachedOwnAccessTokenForClientAuth = fetchOwnAccessTokenResult.token;
+      } else {
+        return { frontendRequest, frontendResponse: fetchOwnAccessTokenResult.errorResponse };
+      }
+
+      // Second try at validation of client-provided token from configured OAuth server
+      verificationResponse = await fetchClientTokenVerification(clientBearerToken, cachedOwnAccessTokenForClientAuth);
+
+      if (verificationResponse.status === 401) {
+        // Something is wrong with the Meadowlark token - still not authorized to perform client-provided
+        // token validation
+        Logger.error(
+          `${moduleName}.authorize OAuth server responded that Meadowlark own token is not valid even after token refresh. Check Meadowlark OAuth configuration.`,
+          frontendRequest.traceId,
+        );
+
+        const errorResponse: FrontendResponse = { body: '', statusCode: 500 };
+        return { frontendRequest, frontendResponse: errorResponse };
+      }
+    }
+    //
+
+    if (verificationResponse.status === 400) {
+      // Client-provided token is not a well-formed JWT
+      Logger.debug(
+        `${moduleName}.authorize verification response returned 400 - Client-provided token is not a JWT`,
+        frontendRequest.traceId,
+      );
+      const errorResponse: FrontendResponse = { body: '', statusCode: 401 };
+      return { frontendRequest, frontendResponse: errorResponse };
+    }
+
+    if (verificationResponse.status === 200) {
+      if (!verificationResponse.data?.active) {
+        // Client-provided token accepted by OAuth server but not active
+        Logger.debug(`${moduleName}.authorize Client-provided token is inactive`, frontendRequest.traceId);
+        const errorResponse: FrontendResponse = { body: '', statusCode: 401 };
+        return { frontendRequest, frontendResponse: errorResponse };
+      }
+
+      // Return client id and Meadowlark authorization strategy (via roles in JWT) to middleware stack
+      const roles: string[] = verificationResponse.data?.roles || [];
+      const clientInfo: VerifiedClientInfo = {
+        security: {
+          clientId: verificationResponse.data?.client_id ?? 'UNKNOWN',
+          authorizationStrategy: determineAuthStrategyFromRoles(roles),
+        },
+        validateResources: !roles.includes('assessment'),
+      };
+
+      cachedTokensForClients.set(clientBearerToken, clientInfo);
+      return { frontendRequest, frontendResponse: null };
```

Comment:
Yes, this function is longer than I generally like to see, as you hinted in the standup. But it is fairly readable with the clear conditions. You probably have something like 

```
if (verificationResponse.status === 200) {
    return authorizationSuccess(... args ...);
}
```

But I'll admit it might not be worth the effort.